compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/TextFieldDelegate.skiko.kt

```diff
@@ -0,0 +1,158 @@
+    } else if (textLayoutResult.isLeftEdgeTapped(offset)) {
+        val lineNumber = textLayoutResult.value.getLineForOffset(offset)
+        caretOffsetPosition = textLayoutResult.value.getLineStart(lineNumber)
+    } else if (textLayoutResult.isRightEdgeTapped(offset)) {
+        val lineNumber = textLayoutResult.value.getLineForOffset(offset)
+        caretOffsetPosition = textLayoutResult.value.getLineEnd(lineNumber)
+    } else if (textLayoutResult.isPunctuationOrSpaceBeforeWordTapped(offset, currentText)) {
+        val nextWordBoundary = findNextWordBoundary(offset, currentText, textLayoutResult)
+        caretOffsetPosition = nextWordBoundary.start
+    } else if (textLayoutResult.isFirstLetterOfWordTapped(offset)) {
+        val wordBoundary = textLayoutResult.value.getWordBoundary(offset)
+        caretOffsetPosition = wordBoundary.start
+    } else {
+        val wordBoundary = textLayoutResult.value.getWordBoundary(offset)
+        caretOffsetPosition = wordBoundary.end
+    }
+
+    onValueChange(editProcessor.toTextFieldValue().copy(selection = TextRange(caretOffsetPosition)))
+}
+
+private fun TextLayoutResultProxy.isPunctuationOrSpaceBeforeWordTapped(
+    caretOffset: Int,
+    currentText: String
+): Boolean {
+    /* From TextLayoutResultProxy.value.getWordBoundary(caretOffset) it is clear
+    * that for whitespace or punctuation mark method will return empty range.
+    * Thus, if empty range was returned, and the caretOffset greater than start and less than last index of text
+    * then offset should be somewhere between words.
+    * */
+    val char = currentText[caretOffset]
+    val isGreaterThanFirst = caretOffset >= 0
+    val isLessThanLast = caretOffset <= currentText.lastIndex
+    return char.isPunctuationOrWhitespace() && isGreaterThanFirst && isLessThanLast
+}
+
+private fun TextLayoutResultProxy.isFirstLetterOfWordTapped(caretOffset: Int): Boolean {
+    return value.getWordBoundary(caretOffset).start == caretOffset
+}
+
+private fun TextLayoutResultProxy.isCaretTapped(
+    caretOffset: Int,
+    previousCaretOffset: Int
+): Boolean {
+    return previousCaretOffset == caretOffset
+}
+
+private fun TextLayoutResultProxy.isLeftEdgeTapped(caretOffset: Int): Boolean {
+    val lineNumber = value.getLineForOffset(caretOffset)
+    val lineStartOffset = value.getLineStart(lineNumber)
+    return lineStartOffset == caretOffset
+}
+
+private fun TextLayoutResultProxy.isRightEdgeTapped(caretOffset: Int): Boolean {
+    val lineNumber = value.getLineForOffset(caretOffset)
+    val lineStartOffset = value.getLineEnd(lineNumber)
+    return lineStartOffset == caretOffset
+}
+
+private fun findNextWordBoundary(
+    caretOffset: Int,
+    currentText: String,
+    textLayoutResult: TextLayoutResultProxy
+): TextRange {
+    val wordRange = textLayoutResult.value.getWordBoundary(caretOffset)
+    val currentChar = currentText[caretOffset]
+    return if (!currentChar.isPunctuationOrWhitespace()) {
```

Comment:
Probably we need to address lower level TODOs instead.
See [1](https://github.com/JetBrains/compose-multiplatform-core/blob/28a2739b41bbf09080067a04bde2e2278fa8643a/compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/SkiaParagraph.skiko.kt#L423-L428), [2](https://github.com/JetBrains/compose-multiplatform-core/blob/28a2739b41bbf09080067a04bde2e2278fa8643a/compose/ui/ui-text/src/skikoTest/kotlin/androidx/compose/ui/text/SkikoParagraphTest.kt#L172)

Not blocking this, but please help to understand why such logic should be on `TextField` level?

Reply:
I've seen that todos and I've thought about that. I decided to keep it here because SkiaParagraph implements Paragraph interface, Paragraph is an expect interface and `getWordBoundaries` should work a bit differently than this method according to its documentation.
So, that's why it's here.
If you have suggestions about it, I'll be glad to hear them.

Reply:
Could add a comment describing the difference between `findNextWordBoundary` and `paragraph.getWordBoundary`?

Reply:
Should I add it here or make a comment in code above this method?

Reply:
In code above the method. It seems that the difference is that we skip spaces and punctuation, but I am not sure

Reply:
done